### PR TITLE
chore: update prettier and migrate to an esm config

### LINKED ISSUE #1501: update prettier
### Is your feature request related to a problem? Please describe.

Prettier 3.0.0 has been released, we should upgrade and migrate to an esm config as its now supported

### Describe the solution you'd like to see

update prettier and alter the current config

### Describe alternate solutions

dont

### Additional information

We can do one pr for the cli and one for the monorepo

### PR DESCRIPTION
**Pull request by FinnDore**
Closes #1501

## ✅ Checklist

- [ ] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [ ] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I performed a functional test on my final commit

---

## Changelog

Updated prettier and converted the cjs config to esm

---

## Screenshots

_[Screenshots]_

image

No screen shot, but a cry for help. each time i install its taking 20mins + ( im still testing my changes )

### CODE REVIEW

File: cli/template/extras/config/prettier.config.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { type Config } from "prettier";
+
+export default {
+  plugins: [require.resolve("prettier-plugin-tailwindcss")],
+} satisfies Config;
```
Comment (juliusmarminge) [inline cli/template/extras/config/prettier.config.ts:5]: Oh shit didnt know this was possible. Lets gooo

File: cli/template/extras/config/prettier.config.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { type Config } from "prettier";
+
+export default {
+  plugins: [require.resolve("prettier-plugin-tailwindcss")],
```
Comment (juliusmarminge) [inline cli/template/extras/config/prettier.config.ts:4]: Why do we have this require.resolve here? Just the string works

File: cli/template/extras/config/prettier.config.ts (hunk 1 above)
Comment (FinnDore) [reply cli/template/extras/config/prettier.config.ts:5]: 😎

Comment (FinnDore) [reply cli/template/extras/config/prettier.config.ts:5]: ive removed the `require.resolve` afaik that was only useful if the config was not at the root of your monorepo? correct me if im wrong.
it seems to work

File: cli/template/extras/config/prettier.config.ts (hunk 2 above)
Comment (FinnDore) [reply cli/template/extras/config/prettier.config.ts]: whops you got to that before me https://github.com/t3-oss/create-t3-app/pull/1500#discussion_r1254026966

File: cli/template/extras/config/prettier.config.ts (hunk 1 above)
Comment (FinnDore) [reply cli/template/extras/config/prettier.config.ts:5]: nvm something isnt working, with or without require. i hope the plugin works with this update

**Comment by FinnDore** [comment]
Looks like this is blocked by https://github.com/tailwindlabs/prettier-plugin-tailwindcss/issues/176

**Comment by juliusmarminge** [comment]
v0.4.0 includes prettier@3 support

**Comment by FinnDore** [comment]
Will get on it 👍

**Comment by FinnDore** [comment]
https://github.com/withastro/prettier-plugin-astro/commit/94ed9046a6d50504e65b58c1fcf505aaa9ac9663
looks like we need to wait for this to be released aswell....

**Comment by FinnDore** [comment]
image
sigh

**Comment by FinnDore** [comment]
Ill confirm not importing the tailwind plugin ( using a string instead ) works for `www` and `upgrade` projects tormorrow. ill also bug some astro people about shipping that new version

**Comment by juliusmarminge** [comment]
> Ill confirm not importing the tailwind plugin ( using a string instead ) works for `www` and `upgrade` projects tormorrow. ill also bug some astro people about shipping that new version

Always Astro huh

File: cli/src/installers/dependencyVersionMap.ts [hunk 3]
```diff
@@ -15,9 +15,9 @@ export const dependencyVersionMap = {
   tailwindcss: "^3.3.0",
   autoprefixer: "^10.4.14",
   postcss: "^8.4.21",
-  prettier: "^2.8.8",
-  "prettier-plugin-tailwindcss": "^0.2.8",
-  "@types/prettier": "^2.7.2",
+  prettier: "^3.0.0",
+  "prettier-plugin-tailwindcss": "^0.4.0",
+  "@types/prettier": "^2.7.3",
```
Comment (Arechii) [inline cli/src/installers/dependencyVersionMap.ts:20]: @types/prettier is no longer needed with v3

Comment (FinnDore) [reply cli/src/installers/dependencyVersionMap.ts:20]: ahh cheers 🙏

**Comment by FinnDore** [comment]
hmm updated the import order sorter and i think its now formatting them altho ive ran into yet another error..... todo with symlinks??? fml
image

**Comment by FinnDore** [comment]
looks like
bad:
```json
    "format": "turbo format && prettier --write *.{cjs,json,md} --ignore-unknown --no-error-on-unmatched-pattern",
```
good:
```json
    "format": "turbo format && prettier --write *.{cjs,json} --ignore-unknown --no-error-on-unmatched-pattern",
```
not sure why, makes no sense to me as to why the md file is causing errors

**Comment by FinnDore** [comment]
ill leave it for now. we can update ivanvs later

**Comment by FinnDore** [comment]
oh the readme is a symlink wtf lol

**Comment by FinnDore** [comment]
the cli project formats it anyways so ill disable that

**Comment by FinnDore** [comment]
image
and we have another

File: cli/prettier.config.mjs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import baseConfig from "../prettier.config.mjs";
+
+/** @type {import('prettier').Config} */
+const config = {
+  ...baseConfig,
+  arrowParens: "always",
+  plugins: [
+    await import("@ianvs/prettier-plugin-sort-imports"),
```
Comment (FinnDore) [inline cli/prettier.config.mjs:8]: https://github.com/t3-oss/create-t3-app/pull/1500#issuecomment-1634749628
this fixes this. im none the wiser

Comment (FinnDore) [reply cli/prettier.config.mjs:8]: not convinced the upgrade project is getting its imports sorted lol

**Comment by FinnDore** [comment]
yeah so its just this im not happy about, other than that this is done
https://github.com/t3-oss/create-t3-app/pull/1500#discussion_r1262969364

**Comment by FinnDore** [comment]
image
oh ffs i just saw this

**Comment by juliusmarminge** [comment]
> image
> 
> oh ffs i just saw this

Might have to bump some eslint versions.

If we're using the prettier api anywhere we also need to make sure we await the result since they ditched the synchronous methods in v3

**Comment by FinnDore** [comment]
ty, sounds like a  job for tomorrow finn

File: upgrade/package.json [hunk 5]
```diff
@@ -1,63 +1,79 @@
 {
-  "name": "@ct3a/upgrade",
```
Comment (FinnDore) [inline upgrade/package.json:2]: what the fuck?

**Comment by FinnDore** [comment]
just pulled the offending files out into a separate repo and it works fine....

**Comment by FinnDore** [comment]
OK! its 99% there. Seems like eslint is borked again. will see whats up with that tomorrow. then i need to clear up this pr, there probbally a bunch of small things + i want to make the prettier configs renames instead of aditions and deletions

**Comment by FinnDore** [comment]
like wtf

https://github.com/t3-oss/create-t3-app/assets/34718806/6e3d9a6e-27d5-4b27-8e0a-a92e6afa38c5

**Comment by FinnDore** [comment]
no way.... i think ive found this issue.... fml

**Comment by FinnDore** [comment]
https://github.com/t3-oss/create-t3-app/assets/34718806/dbb2ab8d-d75c-4d7b-9aa8-7293d2a82f45

**Comment by FinnDore** [comment]
Ok were good here for a review. Should we do the formatting in a new pr ( import order has been updated and has some changes )

**Comment by FinnDore** [comment]
bump :)

File: package.json [hunk 6]
```diff
@@ -51,18 +51,17 @@
     "@total-typescript/ts-reset": "^0.3.7",
     "@types/eslint": "^8.37.0",
     "@types/node": "^18.16.0",
-    "@types/prettier": "^2.7.2",
     "@typescript-eslint/eslint-plugin": "6.0.0",
     "@typescript-eslint/parser": "6.0.0",
     "eslint": "^8.40.0",
     "eslint-config-prettier": "^8.8.0",
```
Comment (juliusmarminge) [inline package.json:57]: i believe there is a v9 available for this, not sure if it's related to prettier v3 though

File: www/prettier.config.mjs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * @type {import('prettier').Config & import('prettier-plugin-tailwindcss').options &
+ *       import("@ianvs/prettier-plugin-sort-imports").PluginConfig}
+ */
+const config = {
+  ...baseConfig,
+  plugins: [
+    "@ianvs/prettier-plugin-sort-imports",
+    "prettier-plugin-astro",
```
Comment (juliusmarminge) [inline www/prettier.config.mjs:11]: is the astro support there yet?

**Review by juliusmarminge — APPROVED** [review]

**Comment by juliusmarminge** [comment]
Some CI failing but rest looks good!

File: package.json (hunk 6 above)
Comment (FinnDore) [reply package.json]: yup, your right, bumped that

File: www/prettier.config.mjs (hunk 7 above)
Comment (FinnDore) [reply www/prettier.config.mjs:11]: yup! https://github.com/withastro/prettier-plugin-astro/releases/tag/v0.11.0
https://github.com/FinnDore/create-t3-app/blob/e5f52c9731d89df768eb67aaa623116c911ff6f6/www/package.json#L57

**Comment by FinnDore** [comment]
the Eslint job is failing because of formatting and the prettier job is also failing because of formatting, and i think upgrade is red cus perms?

Ill create a diff so you can object to any of the formatting changes, i was thinking we could make a new pr for the formatting so this one isnt bloated. wdyt?

**Comment by FinnDore** [comment]
https://github.com/FinnDore/create-t3-app/compare/update-prettier..update-prettier-formatting

**Comment by FinnDore** [comment]
image
im about to scream lmao
image

**Comment by FinnDore** [comment]
ill see if i cant get a minimal repro and make an on ianVs's repo

**Comment by FinnDore** [comment]
https://github.com/t3-oss/create-t3-app/assets/34718806/a74543b6-5b75-41f7-b930-c5ed38c1823f
Im not going insane your going insane,

if figured out the issue tho which is good

**Comment by FinnDore** [comment]
ok it turns out files weren't being formatted because the glob was not quoted, maybe a zsh issue or somthing? either way i think were good now. Although ill have a final check tomorrow

**Comment by juliusmarminge** [comment]
> Ill create a diff so you can object to any of the formatting changes, i was thinking we could make a new pr for the formatting so this one isnt bloated. wdyt?

If you wanna make a PR into this branch from that diff we can review them separetely but still merge to main at the same time?

**Comment by FinnDore** [comment]
Cool, ive tested it again, working as expected, ive also bumped the tailwind plugin

> If you wanna make a PR into this branch from that diff we can review them separetely but still merge to main at the same time?

ye

**Comment by FinnDore** [comment]
heres the link.
https://github.com/FinnDore/create-t3-app/pull/1/files

**Review by juliusmarminge — APPROVED** [review]

**Comment by FinnDore** [comment]
it will take months to recover from this 😔

**Pull request merged**